extensions/ipynb/src/notebookImagePaste.ts

```diff
@@ -0,0 +1,146 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+
+class CopyPasteEditProvider implements vscode.DocumentPasteEditProvider {
+
+	async provideDocumentPasteEdits(
+		_document: vscode.TextDocument,
+		_ranges: readonly vscode.Range[],
+		dataTransfer: vscode.DataTransfer,
+		_token: vscode.CancellationToken
+	): Promise<vscode.DocumentPasteEdit | undefined> {
+
+		// get b64 data from paste
+		// TODO: dataTransfer.get() limits to one image pasted
+		const dataItem = dataTransfer.get('image/png');
+		if (!dataItem) {
+			return undefined;
+		}
+		const fileDataAsUint8 = await dataItem.asFile()?.data();
+		if (!fileDataAsUint8) {
+			return undefined;
+		}
+		const b64string = encodeBase64(fileDataAsUint8);
+
+		// get filename data from paste
+		let pasteFilename = dataItem.asFile()?.name;
+		if (!pasteFilename) {
+			return undefined;
+		}
+		const separatorIndex = pasteFilename?.lastIndexOf('.');
+		const filename = pasteFilename?.slice(0, separatorIndex);
+		const filetype = pasteFilename?.slice(separatorIndex);
+		if (!filename || !filetype) {
+			return undefined;
+		}
+
+		// get notebook cell data
+		let currentCell;
+		for (const notebook of vscode.workspace.notebookDocuments) {
+			if (notebook.uri.path === _document.uri.path) {
+				for (const cell of notebook.getCells()) {
+					if (cell.document === _document) {
+						currentCell = cell;
+						break;
+					}
+				}
+
+			}
+		}
+		if (!currentCell) {
+			return undefined;
+		}
+		const initialAttachments = currentCell.metadata?.custom?.attachments;
+		const initialize = !initialAttachments ? true : false;
+
+		// create updated metadata for cell (prep for WorkspaceEdit)
+		if (!initialize) {
+			let appendValue = 2;
+			while (true) {
+				if (pasteFilename in initialAttachments) {
+					const objEntries = Object.entries(initialAttachments[pasteFilename]);
+					if (objEntries.length) { // check that mime:b64 are present
+						const [, attachmentb64] = objEntries[0];
+						if (attachmentb64 !== b64string) {	// append a "-#" here. same name, diff data. this matches jupyter behavior
+							pasteFilename = filename.concat(`-${appendValue++}`) + filetype;
+							continue;
+						}
+					}
+				}
+				break;
+			}
+		}
+		if (!pasteFilename) {
+			return undefined;
+		}
+		const updatedMetadata = { 'custom': { 'attachments': { [pasteFilename]: { 'image/png': b64string } } } };
+		if (!initialize) {
+			for (const entry of Object.keys(initialAttachments)) {
+				updatedMetadata.custom.attachments[entry] = initialAttachments[entry];
+			}
+		}
+
+		// create WorkspaceEdit and apply to active notebook
+		const newCellMetadata = vscode.NotebookEdit.updateCellMetadata(currentCell.index, updatedMetadata);
+		const edit = new vscode.WorkspaceEdit();
+		if (newCellMetadata) {
+			edit.set(vscode.Uri.file(_document.uri.path), [newCellMetadata]);
```

Comment:
Can you just use `_document.uri` here instead of parsing it again?

Reply:
Unfortunately no, I suspect it's because the uri of `_document` has the `scheme = vscode-notebook-cell`, rather than `file`. The other reason could be the `fragment` property that `_document` has, but I'm not sure the function of that property.

<img width="573" alt="image" src="https://user-images.githubusercontent.com/12552271/182939628-01e491f5-e299-4471-8981-9daa5425929f.png">

Reply:
Ah ok. This is actually unsafe then since it assumes the notebooks is using the `file` scheme. Instead you should use the notebook document uri here